src/full/Agda/Interaction/BuildLibrary.hs

```diff
@@ -0,0 +1,77 @@
+-- | Type-check all files of a library (option @--build-library@).
+
+module Agda.Interaction.BuildLibrary (buildLibrary) where
+
+import           Control.Monad.Except             (throwError)
+import           Control.Monad.IO.Class           (liftIO)
+
+import qualified Data.Set as Set
+
+import           System.Directory                 (getCurrentDirectory)
+import           System.FilePath                  ( (</>) )
+import qualified System.FilePath.Find             as Find
+
+import           Agda.Interaction.FindFile        (hasAgdaExtension)
+import qualified Agda.Interaction.Imports         as Imp
+import           Agda.Interaction.Library         (pattern AgdaLibFile, _libIncludes, getAgdaLibFile)
+import           Agda.Interaction.Options         (optOnlyScopeChecking)
+
+import           Agda.TypeChecking.Monad
+import           Agda.TypeChecking.Pretty         (prettyTCM, text, vsep)
+import           Agda.TypeChecking.Pretty.Warning (getAllWarnings)
+import           Agda.TypeChecking.Warnings       (pattern AllWarnings, classifyWarnings)
+
+import           Agda.Utils.FileName              (absolute)
+import           Agda.Utils.Functor               ()
+import           Agda.Utils.IO.Directory          (findWithInfo)
+import           Agda.Utils.Monad                 (forM, forM_, unless)
+import           Agda.Utils.Null                  (unlessNullM)
+import           Agda.Utils.String                (delimiter)
+
+import           Agda.Utils.Impossible            (__IMPOSSIBLE__)
+
+-- | Find @.agda-lib@ file from current directory
+--   and build all modules located in the @include@ paths
+--   and their subdirectories of the library.
+--
+buildLibrary :: TCM ()
+buildLibrary = do
+  cwd <- liftIO getCurrentDirectory
+
+  -- Read the library file.
+  ls <- libToTCM $ getAgdaLibFile cwd
+  libFile@AgdaLibFile{ _libIncludes = paths } <- case ls of
+    [l] -> pure l
+    []  -> throwError $ GenericException "No library found to build"
+    _   -> __IMPOSSIBLE__
+
+  -- Find all modules in the include paths of the library.
+  files <- map Find.infoPath . concat <$> forM paths \ path -> do
+    liftIO $ findWithInfo (pure True) (hasAgdaExtension <$> Find.filePath) path
+
+  -- Call the type-checker on all these modules.
+  -- (Code copied from Agda.Main.)
+
+  opts <- commandLineOptions
+  let mode = if optOnlyScopeChecking opts
+             then Imp.ScopeCheck
+             else Imp.TypeCheck
+
+  forM_ files \ inputFile -> do
+
+    -- TODO: isolated check
+    -- typeCheckMain does not isolate
+    src <- srcFromPath =<< liftIO (absolute inputFile)
+    result <- Imp.typeCheckMain mode =<< Imp.parseSource src
```

Comment:
To do this properly, we need to factor out a function for isolated checking from `Interaction.Imports`.

Reply:
What do you mean by isolation here? It's important that it's not too isolated, specifically interfaces need to be cached between checks.

Reply:
If you currently run it in `cubical`, it will at some point complain about a duplicate `\ell : Level`, even though both (file) modules declare it `private`.  So it seems like without extra work, the signature (and scope) in `TCState` is not refreshed by `typeCheckMain` and definitions are simply accumulated.

Reply:
Some of the insulation happens in the scope checker (of course). I pushed a commit that adds that, which lets me do `agda --build-library --cubical --guardedness` in cubical. We should add something to disable `InfectiveImport` for the first layer of modules, but otherwise it looks good.